libminifi/include/core/Annotation.h

```diff
@@ -0,0 +1,79 @@
+/**
+
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace core {
+namespace annotation {
+
+template <typename T>
+struct Annotation {
+  const T value_;
```

Comment:
This implicitly disables move, because it's a non-static const data member, inherited by the . It should be `static constexpr` IMO. It also pollutes the scope of derived classes with an overly generic name `value_`, because it's all used through inheritance.

I don't like using inheritance for this purpose. I would prefer if classes could just declare their requirements by creating a static const member or a type alias of a well known name and type, like `struct MyProcessor : Processor { using input_requirements = annotation::input::required; };`.

Reply:
compile-time would be a nice touch, but I don't see how could we verify at compile-time that a processor requiring inputs actually gets some input connections, also wouldn't this require the code to a) either cast all processors to their most derived type b) reimplement the input requirement validation for all processor types?

Reply:
agree with the member pollution concern, we could get rid of the `value_` altogether by trying successive `dynamic_cast`s to `Required`, `Allowed` and `Forbidden`

Reply:
Good point about the compile-time requirements, I didn't think about not having the static type being the most derived type. As a runtime-friendly alternative, we could avoid using inheritance by introducing a new virtual function, something along the lines of `getInputRequirements() const`, or passing the enum value to the `Processor` base constructor for storage.